Modules/_abc.c

```diff
@@ -0,0 +1,1049 @@
+}
+
+static int
+_add_to_registry(PyObject *self, PyObject *cls)
+{
+    _abc_data *impl;
+    int res;
+    impl = _get_impl(self);
+    if (!impl) {
+        return -1;
+    }
+    res = _add_to_weak_set((PyObject *)(impl->_abc_registry), cls, 1);
+    Py_DECREF(impl);
+    return res;
+}
+
+static int
+_add_to_cache(PyObject *self, PyObject *cls)
+{
+    _abc_data *impl;
+    int res;
+    impl = _get_impl(self);
+    if (!impl) {
+        return -1;
+    }
+    res = _add_to_weak_set(impl->_abc_cache, cls, 0);
+    Py_DECREF(impl);
+    return res;
+}
+
+static int
+_add_to_negative_cache(PyObject *self, PyObject *cls)
+{
+    _abc_data *impl;
+    int res;
+    impl = _get_impl(self);
+    if (!impl) {
+        return -1;
+    }
+    res = _add_to_weak_set(impl->_abc_negative_cache, cls, 0);
+    Py_DECREF(impl);
+    return res;
+}
+
+PyDoc_STRVAR(_reset_registry_doc,
+"Internal ABC helper to reset registry of a given class.\n\
+\n\
+Should be only used by refleak.py");
+
+static PyObject *
+_reset_registry(PyObject *m, PyObject *args)
+{
+    PyObject *self, *registry;
+    if (!PyArg_UnpackTuple(args, "_reset_registry", 1, 1, &self)) {
+        return NULL;
+    }
+    registry = _get_registry(self);
+    if (!registry) {
+        return NULL;
+    }
+    if (PySet_Clear(registry) < 0) {
+        return NULL;
+    }
+    Py_RETURN_NONE;
+}
+
+static int
+_reset_negative_cache(PyObject *self)
+{
+    _abc_data *impl;
+    impl = _get_impl(self);
+    if (!impl) {
+        return -1;
+    }
+    if (PySet_Clear(impl->_abc_negative_cache) < 0) {
+        Py_DECREF(impl);
+        return -1;
+    }
+    Py_DECREF(impl);
+    return 0;
+}
+
+PyDoc_STRVAR(_reset_caches_doc,
+"Internal ABC helper to reset both caches of a given class.\n\
+\n\
+Should be only used by refleak.py");
+
+static PyObject *
+_reset_caches(PyObject *m, PyObject *args)
+{
+    PyObject *self;
+    _abc_data *impl;
+    if (!PyArg_UnpackTuple(args, "_reset_caches", 1, 1, &self)) {
+        return NULL;
+    }
+    impl = _get_impl(self);
+    if (!impl) {
+        return NULL;
+    }
+    if (PySet_Clear(impl->_abc_cache) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    /* also the second cache */
+    if (_reset_negative_cache(self) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+    Py_RETURN_NONE;
+}
+
+static PyObject *
+_get_negative_cache_version(PyObject *self)
+{
+    _abc_data *impl;
+    PyObject *res;
+    impl = _get_impl(self);
+    if (!impl) {
+        return NULL;
+    }
+    res = impl->_abc_negative_cache_version;
+    Py_DECREF(impl);
+    return res;
+}
+
+static int
+_set_negative_cache_version(PyObject *self, PyObject *version)
+{
+    _abc_data *impl;
+    impl = _get_impl(self);
+    if (!impl) {
+        return -1;
+    }
+    Py_DECREF(impl->_abc_negative_cache_version);
+    impl->_abc_negative_cache_version = abc_invalidation_counter;
+    Py_INCREF(impl->_abc_negative_cache_version);
```

Comment:
The usual style is to INCREF first.  In this specific case I'd suggest to use `Py_SETREF`:

```C
Py_SETREF(impl->_abc_negative_cache_version, abc_invalidation_counter);
```